torch/csrc/jit/passes/symbolic_shape_cache.cpp

```diff
@@ -0,0 +1,214 @@
+          for (const IValue& elem_ival : ival->toListRef()) {
+            cur_arg = at::hash_combine(cur_arg, IValue::hash(elem_ival));
+          }
+        } else {
+          cur_arg = IValue::hash(ival);
+        }
+      } else {
+        cur_arg = c10::get<CanonicalizedSymbolicShape>(arg).hash();
+      }
+      hash_val = at::hash_combine(hash_val, cur_arg);
+    }
+    return hash_val;
+  }
+};
+
+using ShapeCache = lazy::Cache<
+    ShapeCacheKey,
+    std::vector<CanonicalizedSymbolicShape>,
+    ArgumentsHasher>;
+
+constexpr size_t kShapeCacheSize = 1024;
+ShapeCache shapeCache(kShapeCacheSize);
+
+ShapeCacheKey get_cache_key(
+    const FunctionSchema* schema,
+    const std::vector<SSAInput>& arg_vec,
+    std::unordered_map<int64_t, int64_t>& ss_map,
+    bool deep_copy = true) {
+  CanonicalArgVec canonical_args = cannonicalizeVec(arg_vec, ss_map, deep_copy);
+  return std::make_tuple(schema->operator_name(), canonical_args);
+}
+
+} // namespace
+
+TORCH_API void cache_shape_function(
+    const FunctionSchema* schema,
+    const std::vector<SSAInput>& arg_vec,
+    const std::vector<at::SymbolicShape>& ret_vec) {
+  // TODO: compare perf using std::vector<std::tuple<int64_t, int64_t>>
+  auto ss_map = std::unordered_map<int64_t, int64_t>();
+  auto cache_key = get_cache_key(schema, arg_vec, ss_map, /* deep_copy */ true);
+  auto can_ret_vec = std::make_shared<std::vector<CanonicalizedSymbolicShape>>(
+      cannonicalizeVec(ret_vec, ss_map));
+  shapeCache.Add(cache_key, can_ret_vec);
+}
+
+TORCH_API c10::optional<std::vector<at::SymbolicShape>>
+get_cached_shape_function(
+    const FunctionSchema* schema,
+    const std::vector<SSAInput>& arg_vec) {
+  // TODO: compare perf using std::vector<std::tuple<int64_t, int64_t>> for both
+  // ss_map and inverse_ss_map
+  auto ss_map = std::unordered_map<int64_t, int64_t>();
+  auto cache_key =
+      get_cache_key(schema, arg_vec, ss_map, /* deep_copy */ false);
+  auto cached_ret_vec = shapeCache.Get(cache_key);
+  if (cached_ret_vec == nullptr) {
+    return c10::nullopt;
+  }
+  // Decanonicalize the return values
+  auto inverse_ss_map = std::unordered_map<int64_t, int64_t>();
+  for (auto& ss_val : ss_map) {
+    inverse_ss_map[ss_val.second] = ss_val.first;
+  }
+  std::vector<at::SymbolicShape> ret_vec;
+  for (auto& css : *cached_ret_vec) {
+    ret_vec.emplace_back(css.toSymbolicShape(inverse_ss_map));
+  }
+  return ret_vec;
+}
+
+// Function only to access the cache, used for testing
+TORCH_API void clear_shape_cache() {
+  shapeCache.Clear();
+}
+
+TORCH_API size_t get_shape_cache_size() {
+  return shapeCache.Numel();
+}
+
+void CanonicalizedSymbolicShape::init(
+    const c10::SymbolicShape& orig_shape,
+    std::unordered_map<int64_t, int64_t>& ss_map) {
+  auto sizes = orig_shape.sizes();
+  if (!sizes) {
+    values_ = c10::nullopt;
+    return;
+  }
+  values_ = std::vector<int64_t>();
+  int64_t cur_symbolic_index = -(int64_t)ss_map.size() - 1;
```

Comment:
nit: (no need to update if this is only change but) generally better practice to use static_cast afaik